hledger-lib/Hledger/Data/Account.hs

```diff
@@ -64,7 +64,7 @@ accountsFromPostings :: [Posting] -> [Account]
 accountsFromPostings ps =
   let
     acctamts = [(paccount p,pamount p) | p <- ps]
-    grouped = groupBy (\a b -> fst a == fst b) $ sort $ acctamts
+    grouped = groupBy (\a b -> fst a == fst b) $ sortBy (\a b -> compare (fst a) (fst b)) $ acctamts
```

Comment:
Whole expression sounds like a `groupSortOn` from [`Data.List.Extra` of "`extra` package](https://hackage.haskell.org/package/extra-1.6/docs/Data-List-Extra.html).

Reply:
This would internally generate (accountname, (accountname, amount)) list, which sounds a bit silly and might outweight the cost of adding one extra package to the build ... 

Shortening the code to `groupBy ((==) `on` fst) . sortBy (compare `on` fst)` sounds nice, though

Reply:
The reason why it was written like that is to re-use computations made at sort. Though extra thumbs shouldn't introduce much of performance degradation.

If it have significant impact and we are ok to trade readability we can use:

```hledger
grouped = groupSortBy (compare `on` fst)
```

This implementation assumes that comparison is not very complex and can be re-computed in a similar way we do here.

Reply:
I might be a bit dense here ... how exactly does it re-use computations made at sort?

Reply:
```haskell
groupSortOn f = map (map snd) . groupBy ((==) `on` fst) . sortBy (compare `on` fst) . map (f &&& id)
```
With this approach we calculate `f` only once when we require value during `compare` within `sortBy`. Alternative approach would be:

```haskell
groupSortOn f = groupBy ((==) `on` f) . sortBy (compare `on` f)
```

But then it will assume that `f` is lighter than introducing new tuple. And there is no way to optimize it outside without use of memoization.

That's how I justify why `groupSortOn` introduce temporary tuple which actually can be dropped by compiler.

Reply:
Maybe we are talking across each other. Are you proposing to replace this:
```
acctamts = [(paccount p,pamount p) | p <- ps]
grouped = groupBy ((==) `on` fst) $ sortBy (compare `on` fst) $ acctamts
counted = [(a, length acctamts) | acctamts@((a,_):_) <- grouped]
summed = map (\as@((aname,_):_) -> (aname, sumStrict $ map snd as)) grouped -- always non-empty
```

With this: 
```
grouped =  map (\grp -> (paccount (head gpr), map pamount grp)) $ groupSortOn paccount ps
counted = [(a, length amts) | (a, amts) <- grouped]
summed = map (\(aname,as) -> (aname, sumStrict $ map snd as)) grouped -- always non-empty
```
?

This would make sense, but i will still maintain that just writing `grouped = groupSortOn fst acctamts` (where `acctamts :: (AccountName, Amount)`) would just introduce overhead of extra package and extra tuple-of-tuple built for no real benefit.

Reply:
@adept , no I don't suggest to extend this change to many lines of code. I like small and precise hit :)

I just tried to suggest a ways to render your idea with hiding internals outside of that line of code as much as possible.

If you'll do `git grep 'group.*sort' ':/**/*.hs'` you'll find other place where this pattern of `group . sort` repeats. I just felt that someone already had to implement that `groupSort`-kind of functions.

I didn't considered that adding of package `extra` may introduce overhead.

Reply:
Omg, I got `groupSortOn` and `groupSortBy` horribly mixed up, thank you for being patient with me and not calling me an idiot I certainly was. I'll push more changes later

Reply:
@adept I think it is easy to mix `On` with `By`.
Now when I understand your last snipped of code it looks like `groupSort` is actually what we want here.

```haskell
groupSort :: Ord k => [(k, v)] -> [(k, [v])]
groupSort = map (\x -> (fst $ head x, map snd x)) . groupOn fst . sortOn fst
```
I.e.
```haskell
grouped =  map (\grp -> (paccount (head gpr), map pamount grp)) $ groupSortOn paccount ps
```
Can be written as
```haskell
grouped = groupSort $ map (paccount &&& pamount) ps
```